Source/JavaScriptCore/jit/ExecutableAllocator.h

```diff
@@ -119,6 +119,71 @@ JS_EXPORT_PRIVATE void dumpJITMemory(const void*, const void*, size_t);
 JS_EXPORT_PRIVATE void* performJITMemcpyWithMProtect(void *dst, const void *src, size_t n);
 #endif
 
+#if ENABLE(JIT_SCAN_ASSEMBLER_BUFFER_FOR_ZEROES)
+// This helps with error logging in the case of a crash, as it places the crash
+// PC at the point in the instruction stream which would be corrupted --
+// while preserving the stack so that we can use that to figure out who might be
+// responsible for generating the corrupted instructions.
+template<typename T, typename U, typename V>
+static NEVER_INLINE NO_RETURN_DUE_TO_CRASH NOT_TAIL_CALLED void dieByJumpingIntoJITBufferWithInfo(int line, void* buffer, size_t offset, size_t size, T info1, U info2, V info3)
+{
+    RELEASE_ASSERT(offset <= size);
+    RELEASE_ASSERT(offset <= std::numeric_limits<uint32_t>::max());
+    RELEASE_ASSERT(size <= std::numeric_limits<uint32_t>::max());
+    void* targetInstr = reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(buffer) + offset);
+
+    // Since this function should only be called w/ buffer+offset pointing at
+    // already-zeroed memory, re-zeroing it is somewhat superflous on ARM64E,
+    // but necessary on x86-64 (where 0x00 0x00 encodes addb %al, (%rax)), and
+    // even on the former ensures that execution will never continue past the
+    // branch out of this function even if it is called improperly.
+#if OS(DARWIN) && CPU(X86_64)
+    memset(reinterpret_cast<char*>(targetInstr), 0xF4, 1);
+    sys_icache_invalidate(buffer, size);
+#elif OS(DARWIN) && CPU(ARM64)
+    memset(reinterpret_cast<char*>(targetInstr), 0, 4);
+    sys_icache_invalidate(buffer, size);
+#else
+#error "JIT_SCAN_ASSEMBLER_BUFFER_FOR_ZEROES not supported on this platform."
+#endif
+
+    register uint64_t lineGpr asm(CRASH_GPR0) = static_cast<uint64_t>(static_cast<int64_t>(line));
+    register uint64_t bufferGpr asm(CRASH_GPR1) = reinterpret_cast<uintptr_t>(buffer);
+    register uint64_t sizeOffsetGpr asm(CRASH_GPR2) = static_cast<uint64_t>(offset | size << 32);
+    register uint64_t info3Gpr asm(CRASH_GPR5) = wtfCrashArg(info3);
+    register uint64_t info2Gpr asm(CRASH_GPR4) = wtfCrashArg(info2);
+    register uint64_t info1Gpr asm(CRASH_GPR3) = wtfCrashArg(info1);
+
+#if CPU(X86_64)
+    __asm__ volatile("jmp *%0" : : "r"(targetInstr), "r"(lineGpr), "r"(bufferGpr), "r"(sizeOffsetGpr), "r"(info1Gpr), "r"(info2Gpr), "r"(info3Gpr));
+#elif CPU(ARM64)
+    __asm__ volatile("br %0" : : "r"(targetInstr), "r"(lineGpr), "r"(bufferGpr), "r"(sizeOffsetGpr), "r"(info1Gpr), "r"(info2Gpr), "r"(info3Gpr));
+#else
+#error "JIT_SCAN_ASSEMBLER_BUFFER_FOR_ZEROES not supported on this platform."
+#endif
+    __builtin_unreachable();
```

Comment:
Can't we just do 
```
CRASH_WITH_INFO(targetInstr, static_cast<uint64_t>(static_cast<int64_t>(line)), reinterpret_cast<uintptr_t>(buffer), static_cast<uint64_t>(offset | size << 32), wtfCrashArg(info1), wtfCrashArg(info2), wtfCrashArg(info3));
```

Reply:
In that case StabilityTracer would give us the instruction stream of the compiler; we can get a bit more info (i.e. the x86 instructions immediately prior to the ones being corrupted) if we jump into the buffer, essentially imitating what would happen if we didn't catch this corruption during the zero-checking stage. However, since this function is NEVER_INLINE we preserve the prior stack frame and thus get the same stack trace as we were in the zero-checking patch.

Obviously it's a lot of complexity for a bit of diagnostic info, but I couldn't think of any other way to keep that information available.